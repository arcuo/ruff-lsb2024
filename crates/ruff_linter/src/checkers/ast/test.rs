use std::path::Path;

use ruff_python_index::Indexer;
use ruff_python_parser::{lexer::LexResult, Mode};
use ruff_python_semantic::Module;
use ruff_source_file::Locator;

use crate::{
    directives, linter::TokenSource, packaging::detect_package_root, settings::{flags::Noqa, LinterSettings}
};

use super::Checker;

// TODO: Should we use this?

/// Create a new [`Checker`] for the given source code
pub(crate) fn test_checker_state_after_snippet_traversal(source: &str) -> &Checker {
    let path = Path::new("");
    let locator = Locator::new(source);
    let tokens = TokenSource::Tokens(ruff_python_parser::tokenize_all(source, Mode::Module));
    let indexer = Indexer::from_tokens(&tokens, &locator);
    let settings = LinterSettings::default();
    let directives = directives::extract_directives(
        &tokens,
        directives::Flags::from_settings(&settings),
        &locator,
        &indexer,
    );
    let checker = Checker::new(
        &settings,
        &directives.noqa_line_for,
        Noqa::Disabled,
        &path,
        path.parent()
            .and_then(|parent| detect_package_root(parent, &settings.namespace_packages)),
        Module {
            kind: ruff_python_semantic::ModuleKind::Module,
            source: ruff_python_semantic::ModuleSource::Path(&[path.to_str().unwrap().to_string()]),
            python_ast: tokens.,
            name: path.to_string(),
        },
    );
    &checker
}
