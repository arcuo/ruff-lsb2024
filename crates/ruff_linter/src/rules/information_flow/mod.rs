//! Rules for information flow.
pub(crate) mod rules;

#[cfg(test)]
mod tests {
    use std::path::Path;

    use anyhow::Result;
    use test_case::test_case;

    use crate::registry::Rule;
    use crate::test::test_path;
    use crate::{assert_messages, settings};

    // Basic rules
    #[test_case(Rule::IFMustIncludeVariableLabel, Path::new("IF001.py"))]
    #[test_case(Rule::IFMissingPrincipal, Path::new("IF002.py"))]
    // Explicit
    #[test_case(Rule::IFExplicitVariableAssign, Path::new("IF101.py"))] // TODO: split into multiple tests
    #[test_case(Rule::IFExplicitVariableAssign, Path::new("IF101_for.py"))] // TODO: split into multiple tests

    // Implicit
    #[test_case(Rule::IFImplicitVariableAssign, Path::new("IF201_if.py"))]
    #[test_case(Rule::IFImplicitVariableAssign, Path::new("IF201_while.py"))]
    #[test_case(Rule::IFImplicitVariableAssign, Path::new("IF201_for.py"))]
    #[test_case(Rule::IFImplicitVariableAssign, Path::new("IF201_nested.py"))]
    #[test_case(Rule::IFImplicitVariableAssign, Path::new("IF201_assert.py"))]

    fn rules(rule_code: Rule, path: &Path) -> Result<()> {
        let snapshot = format!("{}_{}", rule_code.noqa_code(), path.to_string_lossy());
        let diagnostics = test_path(
            Path::new("information_flow").join(path).as_path(),
            &settings::LinterSettings::for_rule(rule_code),
        )?;
        assert_messages!(snapshot, diagnostics);
        Ok(())
    }
}
