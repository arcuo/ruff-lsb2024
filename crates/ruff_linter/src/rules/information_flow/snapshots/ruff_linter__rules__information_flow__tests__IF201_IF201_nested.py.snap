---
source: crates/ruff_linter/src/rules/information_flow/mod.rs
---
IF201_nested.py:11:5: IF201 Illegal implicit information flow. Current pc label: {secret} is greater than target label: {}. Current pc set at line 10
   |
10 | if secret_var > 0: # PC = {secret}
11 |     public_var = 0 # Fail
   |     ^^^^^^^^^^ IF201
12 |     while public_var < 10: # PC = {secret}
13 |         public_var += 1 # Fail
   |

IF201_nested.py:13:9: IF201 Illegal implicit information flow. Current pc label: {secret} is greater than target label: {}. Current pc set at line 10
   |
11 |     public_var = 0 # Fail
12 |     while public_var < 10: # PC = {secret}
13 |         public_var += 1 # Fail
   |         ^^^^^^^^^^ IF201
14 | 
15 | if public_var > 0: # PC = {}
   |

IF201_nested.py:18:9: IF201 Illegal implicit information flow. Current pc label: {secret} is greater than target label: {}. Current pc set at line 17
   |
16 |     public_var = 0 # Success
17 |     while secret_var < 10: # PC = {secret}
18 |         public_var += 1 # Fail
   |         ^^^^^^^^^^ IF201
19 | 
20 | ## Nested IF in WHILE
   |

IF201_nested.py:23:5: IF201 Illegal implicit information flow. Current pc label: {secret} is greater than target label: {}. Current pc set at line 22
   |
22 | while secret_var < 10: # PC = {secret}
23 |     public_var = 0 # Fail
   |     ^^^^^^^^^^ IF201
24 |     secret_var += 1 # Success
25 |     if public_var < 10: # PC = {secret}
   |

IF201_nested.py:26:9: IF201 Illegal implicit information flow. Current pc label: {secret} is greater than target label: {}. Current pc set at line 22
   |
24 |     secret_var += 1 # Success
25 |     if public_var < 10: # PC = {secret}
26 |         public_var += 1 # Fail
   |         ^^^^^^^^^^ IF201
27 | 
28 | while public_var < 10:
   |

IF201_nested.py:37:1: IF201 Illegal implicit information flow. Current pc label: {secret} is greater than target label: {}. Current pc set at line 35
   |
35 | assert secret_var > 0
36 | 
37 | public_var = 0 # Fail
   | ^^^^^^^^^^ IF201
38 | 
39 | if public_var == 0:
   |

IF201_nested.py:40:5: IF201 Illegal implicit information flow. Current pc label: {secret} is greater than target label: {}. Current pc set at line 35
   |
39 | if public_var == 0:
40 |     public_var += 1 # Fail
   |     ^^^^^^^^^^ IF201
41 | 
42 | while public_var < 10:
   |

IF201_nested.py:43:5: IF201 Illegal implicit information flow. Current pc label: {secret} is greater than target label: {}. Current pc set at line 35
   |
42 | while public_var < 10:
43 |     public_var += 1 # Fail
   |     ^^^^^^^^^^ IF201
   |
